drv/gimlet-hf-server/src/main.rs

```diff
@@ -505,14 +505,21 @@ impl idl::InOrderHostFlashImpl for ServerImpl {
     fn read(
         &mut self,
         _: &RecvMessage,
-        addr: u32,
-        dest: LenLimit<Leased<W, [u8]>, PAGE_SIZE_BYTES>,
+        mut addr: u32,
+        dest: Leased<W, [u8]>,
     ) -> Result<(), RequestError<HfError>> {
         self.check_muxed_to_sp()?;
-        self.qspi.read_memory(addr, &mut self.block[..dest.len()]);
+        let mut offset = 0;
+        for i in (0..dest.len()).step_by(self.block.len()) {
```

Comment:
**suggestion:** Since `i` is only used to calculate `n` which itself is either the block length or smaller than that, would it perhaps be clearer to rewrite this as a `k = dest.len() % block.len()` loop, each reading `block.len()` bytes, followed by a final step of reading the remainder? Alternatively, combine the two and do the min calculation on every loop of course.

I had to wrack my brain a good while to figure out what the `n` calculation was doing.

**suggestion:** Maybe perform `self.block.len()` only once (and `dest.len()` as well?) and reuse the result: I didn't check with Godbolt but I wouldn't be surprised if the mutable access to `dest` and `self.block` wouldn't cause the compiler to think it has to reload the lengths, even though they presumably never do change.

Reply:
**note:** While looking at my folly, I noticed that it seems like `step_by` is stopping some compiler work from happening: https://godbolt.org/z/6Phz5WEcG

It's a shame that such a useful API drops code off a (probably minor) cliff :(

I hope this at least is interesting and potentially useful information to you.